perfkitbenchmarker/vpn_service.py

```diff
@@ -0,0 +1,326 @@
+import time
+import uuid
+
+from perfkitbenchmarker import context, vm_util
+from perfkitbenchmarker import errors
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import resource
+
+flags.DEFINE_integer('vpn_service_tunnel_count', None,
+                     'Number of tunnels to create for each VPN GW pair.')
+flags.DEFINE_integer('vpn_service_gateway_count', None,
+                     'Number of VPN GWs to create for each vm_group.')
+flags.DEFINE_string('vpn_service_name', None,
+                    'If set, use this name for VPN Service.')
+flags.DEFINE_string('vpn_service_shared_key', None,
+                    'If set, use this PSK for VPNs.')
+flags.DEFINE_string('vpn_service_routing_type', None,
+                    'static or dynamic(BGP)')
+flags.DEFINE_integer('vpn_service_ike_version', None, 'IKE version')
+
+FLAGS = flags.FLAGS
+
+
+def GetVPNServiceClass():
+  """Gets the VPNService class.
+
+  Args:
+
+  Returns:
+    Implementation class
+  """
+  return resource.GetResourceClass(VPNService)
+
+
+class VPN(object):
+  """An object representing the VPN.
+
+  A VPN instance manages tunnel configurations for exactly 1 pair of endpoints.
+  """
+
+  def __init__(self, *args, **kwargs):
+      return object.__init__(self, *args, **kwargs)
+
+  def getKeyFromGWPair(self, gateway_pair, suffix=''):
+    """Return the VPN key for a pair of endpoints.
+
+    Args:
+      gateway_pair: A tuple of 2 VPN gateways which define the VPN tunnel.
+      suffix: A unique suffix if multiple tunnels b/t this gateway pair exist.
+
+    Returns:
+      string. The VPN key.
+
+    """
+    key = 'vpn' + ''.join(gw for gw in gateway_pair) + suffix + FLAGS.run_uri
+    return key
+
+  def Create(self, gateway_pair, suffix=''):
+    self.GWPair = gateway_pair
+    self.name = self.getKeyFromGWPair(gateway_pair)
+    self.tunnel_config = TunnelConfig(tunnel_name=self.name, suffix=suffix)
+
+  def Delete(self):
+    pass
+
+  def GetVPN(self, gateway_pair, suffix=''):
+    """Gets a VPN object for the gateway_pair or creates one if none exists
+
+    Args:
+    gateway_pair: a tuple of two VPNGWs
+    """
+
+    benchmark_spec = context.GetThreadBenchmarkSpec()
+    if benchmark_spec is None:
+      raise errors.Error('GetVPN called in a thread without a '
+                         'BenchmarkSpec.')
+    with benchmark_spec.vpns_lock:
+      key = self.getKeyFromGWPair(gateway_pair, suffix)
+      if key not in benchmark_spec.vpns:
+        self.Create(gateway_pair, suffix)
+        benchmark_spec.vpns[key] = self
+      return benchmark_spec.vpns[key]
+
+  def ConfigureTunnel(self):
+    """Configure the VPN tunnel."""
+
+    benchmark_spec = context.GetThreadBenchmarkSpec()
+    vpn_gateway_0 = benchmark_spec.vpn_gateways[self.GWPair[0]]
+    vpn_gateway_1 = benchmark_spec.vpn_gateways[self.GWPair[1]]
+
+    assert not (vpn_gateway_0.require_target_to_init and vpn_gateway_1.require_target_to_init), 'Cant connect 2 passive VPN GWs'
```

Comment:
Can you create a temporary variable that would make this esaier to read?  ie:
  both_require_init = vpn_gateway_0.require_target_to_init and vpn_gateway_1.require_target_to_init
  assert not both_require_init, '....'

Reply:
Fixed.